cylc/flow/taskdef.py

```diff
@@ -73,26 +73,42 @@ def generate_graph_children(tdef, point):
     return graph_children
 
 
-def generate_graph_parents(tdef, point):
-    """Determine graph parents of this task."""
+def generate_graph_parents(tdef, point, taskdefs):
+    """Determine graph parents of this task at this point."""
     graph_parents = {}
     for seq, ups in tdef.graph_parents.items():
+        if not seq.is_valid(point):
+            # Ignore this upstream trigger if the child point is not on
+            # the sequence (recurrence) the trigger belongs to. E.g.:
+            #   PT6H = "waz"
+            #   T06 = "waz[-PT6H] => foo"
+            # Here waz[-PT6H] is a parent of T06/foo but not T12/foo.
+            continue
         graph_parents[seq] = []
-        for name, trigger in ups:
+        for parent_name, trigger in ups:
             parent_point = trigger.get_parent_point(point)
+            if (
+                parent_point != point and
+                not taskdefs[parent_name].is_valid_point(parent_point)
+            ):
+                # Intercycle trigger. Ignore if inferred parent is not valid.
+                # If parent and child have same name, the offset point could
+                # still be pre-initial and should be ignored.
+                #   foo[-Px] => foo
+                # Otherwise it could be pre-initial, OR an erroneous offset
+                # (recurrence membership is not defined by offset triggers).
+                #   woo[-Px] => foo
+                # TODO should we check this elsewhere and raise an error?
+                continue
             is_abs = (trigger.offset_is_absolute or
                       trigger.offset_is_from_icp)
             if is_abs and parent_point != point:
                 # If 'foo[^] => bar' only spawn off of '^'.
                 continue
-            if seq.is_valid(parent_point):
-                # E.g.: foo should trigger only on T06:
-                #   PT6H = "waz"
-                #   T06 = "waz[-PT6H] => foo"
-                graph_parents[seq].append((name, parent_point, is_abs))
+            graph_parents[seq].append((parent_name, parent_point, is_abs))
 
     if tdef.sequential:
-        # Add prev-instance parent.
+        # Add implicit prevous-instance parent.
```

Comment:
```suggestion
        # Add implicit previous-instance parent.
```

(Add with skip-ci once tests all passing)